ddmrp/models/purchase_order.py

```diff
@@ -40,3 +55,31 @@ def _calc_execution_priority(self):
         (self - to_compute).write(
             {"execution_priority_level": None, "on_hand_percent": None}
         )
+
+    def _get_domain_buffer_link(self):
+        self.ensure_one()
+        if not self.product_id:
+            # Return impossible domain -> no buffer.
+            return [(0, "=", 1)]
+        return [
+            ("product_id", "=", self.product_id.id),
+            ("company_id", "=", self.order_id.company_id.id),
+            ("buffer_profile_id.item_type", "=", "purchased"),
+            ("warehouse_id", "=", self.order_id.picking_type_id.warehouse_id.id),
+        ]
+
+    def _find_buffer_link(self):
+        buffer_model = self.env["stock.buffer"]
+        move_model = self.env["stock.move"]
+        for rec in self.filtered(lambda r: not r.buffer_ids):
```

Comment:
This is looping twice on the records: once to filter the recordset, once in the for loop.
It is more efficient to filter in the same iteration as the for loop:

```
for rec in self:
    if not rec.buffer_ids:
        continue
```

(although in practice, you'll never receive more than one record at a time given the way the method is called now)